libsolidity/ast/AST.cpp

```diff
@@ -389,6 +391,49 @@ std::vector<std::pair<ASTPointer<IdentifierPath>, std::optional<Token>>> UsingFo
 	return ranges::zip_view(m_functionsOrLibrary, m_operators) | ranges::to<vector>;
 }
 
+void StructDefinition::insertEip712EncodedSubtypes(std::set<std::string>& subtypes) const
+{
+	for (size_t i = 0; i < m_members.size(); i++)
+	{
+		if (m_members[i]->type()->category() == Type::Category::Struct)
```

Comment:
this ignores the case of arrays of structs `S[..] x` as a member field. I've made correction on my end, not sure if I can push to the 1inch branch:

```diff
diff --git a/libsolidity/ast/AST.cpp b/libsolidity/ast/AST.cpp
index 25d1c1eeb..bb350b4fb 100644
--- a/libsolidity/ast/AST.cpp
+++ b/libsolidity/ast/AST.cpp
@@ -391,19 +391,38 @@ std::vector<std::pair<ASTPointer<IdentifierPath>, std::optional<Token>>> UsingFo
 	return ranges::zip_view(m_functionsOrLibrary, m_operators) | ranges::to<vector>;
 }
 
+
+
 void StructDefinition::insertEip712EncodedSubtypes(std::set<std::string>& subtypes) const
 {
 	for (size_t i = 0; i < m_members.size(); i++)
 	{
-		if (m_members[i]->type()->category() == Type::Category::Struct)
+		Declaration const* declaration;
+
+		switch (m_members[i]->type()->category())
 		{
-			Declaration const* declaration = m_members[i]->type()->typeDefinition();
-			StructDefinition const* structDef = dynamic_cast<StructDefinition const*>(declaration);
-			solAssert(structDef != nullptr);
+			case Type::Category::Struct:
+				declaration = m_members[i]->type()->typeDefinition();
+				break;
+			case Type::Category::Array:
+				{
+					auto const* arrayType = dynamic_cast<ArrayType const*>(m_members[i]->type());
+					solAssert(!!arrayType);
+					auto finalyBaseType = dynamic_cast<StructType const*>(arrayType->finalBaseType(false));
+					solAssert(!!finalyBaseType);
 
-			subtypes.insert(structDef->eip712EncodeTypeWithoutSubtypes());
-			structDef->insertEip712EncodedSubtypes(subtypes);
+					declaration = finalyBaseType->typeDefinition();
+				}
+				break;
+			default:
+				continue;
 		}
+
+		StructDefinition const* structDef = dynamic_cast<StructDefinition const*>(declaration);
+		solAssert(!!structDef);
+
+		subtypes.insert(structDef->eip712EncodeTypeWithoutSubtypes());
+		structDef->insertEip712EncodedSubtypes(subtypes);
 	}
 }
 
@@ -413,7 +432,7 @@ std::string StructDefinition::eip712EncodeTypeWithoutSubtypes() const
 	for (size_t i = 0; i < m_members.size(); i++)
 	{
 		str += i == 0 ? "" : ",";
-		str += m_members[i]->type()->canonicalName() + " " + m_members[i]->name();
+		str += m_members[i]->type()->eip712TypeName() + " " + m_members[i]->name();
 	}
 	return str + ")";
 }
diff --git a/libsolidity/ast/Types.cpp b/libsolidity/ast/Types.cpp
index 5106648b1..5d4a796c5 100644
--- a/libsolidity/ast/Types.cpp
+++ b/libsolidity/ast/Types.cpp
@@ -1887,6 +1887,23 @@ string ArrayType::canonicalName() const
 	return ret;
 }
 
+std::string ArrayType::eip712TypeName() const
+{
+	string ret;
+	if (isString())
+		ret = "string";
+	else if (isByteArrayOrString())
+		ret = "bytes";
+	else
+	{
+		ret = baseType()->eip712TypeName() + "[";
+		if (!isDynamicallySized())
+			ret += length().str();
+		ret += "]";
+	}
+	return ret;
+}
+
 string ArrayType::signatureInExternalFunction(bool _structsByName) const
 {
 	if (isByteArrayOrString())
@@ -2511,6 +2528,11 @@ string StructType::canonicalName() const
 	return *m_struct.annotation().canonicalName;
 }
 
+string StructType::eip712TypeName() const
+{
+	return this->typeDefinition()->name();
+}
+
 FunctionTypePointer StructType::constructorType() const
 {
 	TypePointers paramTypes;
diff --git a/libsolidity/ast/Types.h b/libsolidity/ast/Types.h
index fa1a299a9..10dd75837 100644
--- a/libsolidity/ast/Types.h
+++ b/libsolidity/ast/Types.h
@@ -357,6 +357,7 @@ public:
 	/// @returns the canonical name of this type for use in library function signatures.
 	virtual std::string canonicalName() const { return toString(true); }
 	virtual std::string humanReadableName() const { return toString(); }
+	virtual std::string eip712TypeName() const { return encodingType()->toString(true); }
 	/// @returns the signature of this type in external functions, i.e. `uint256` for integers
 	/// or `(uint256,bytes8)[2]` for an array of structs. If @a _structsByName,
 	/// structs are given by canonical name like `ContractName.StructName[2]`.
@@ -867,6 +868,7 @@ public:
 	std::string toString(bool _withoutDataLocation) const override;
 	std::string humanReadableName() const override;
 	std::string canonicalName() const override;
+	std::string eip712TypeName() const override;
 	std::string signatureInExternalFunction(bool _structsByName) const override;
 	MemberList::MemberMap nativeMembers(ASTNode const* _currentScope) const override;
 	Type const* encodingType() const override;
@@ -924,6 +926,7 @@ public:
 	BoolResult isImplicitlyConvertibleTo(Type const& _other) const override;
 	BoolResult isExplicitlyConvertibleTo(Type const& _convertTo) const override;
 	std::string richIdentifier() const override;
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool operator==(Type const& _other) const override;
 	unsigned calldataEncodedSize(bool) const override { solAssert(false, ""); }
 	unsigned calldataEncodedTailSize() const override { return 32; }
@@ -1052,6 +1055,7 @@ public:
 	std::unique_ptr<ReferenceType> copyForLocation(DataLocation _location, bool _isPointer) const override;
 
 	std::string canonicalName() const override;
+	std::string eip712TypeName() const override;
 	std::string signatureInExternalFunction(bool _structsByName) const override;
 
 	/// @returns a function that performs the type conversion between a list of struct members
@@ -1187,6 +1191,7 @@ public:
 
 	std::string toString(bool _withoutDataLocation) const override;
 	std::string canonicalName() const override;
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	std::string signatureInExternalFunction(bool) const override { solAssert(false, ""); }
 
 protected:
@@ -1218,6 +1223,8 @@ public:
 	bool hasSimpleZeroValueInMemory() const override { return false; }
 	Type const* mobileType() const override;
 
+	std::string eip712TypeName() const override { solAssert(false, ""); }
+
 	std::vector<Type const*> const& components() const { return m_components; }
 
 protected:
@@ -1415,6 +1422,7 @@ public:
 	TypeResult binaryOperatorResult(Token, Type const*) const override;
 	std::string canonicalName() const override;
 	std::string humanReadableName() const override;
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	std::string toString(bool _withoutDataLocation) const override;
 	unsigned calldataEncodedSize(bool _padded) const override;
 	bool canBeStored() const override { return m_kind == Kind::Internal || m_kind == Kind::External; }
@@ -1551,6 +1559,7 @@ public:
 	bool operator==(Type const& _other) const override;
 	std::string toString(bool _withoutDataLocation) const override;
 	std::string canonicalName() const override;
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool containsNestedMapping() const override { return true; }
 	TypeResult binaryOperatorResult(Token, Type const*) const override { return nullptr; }
 	Type const* encodingType() const override;
@@ -1595,6 +1604,7 @@ public:
 	bool operator==(Type const& _other) const override;
 	bool canBeStored() const override { return false; }
 	u256 storageSize() const override;
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool hasSimpleZeroValueInMemory() const override { solAssert(false, ""); }
 	std::string toString(bool _withoutDataLocation) const override { return "type(" + m_actualType->toString(_withoutDataLocation) + ")"; }
 	MemberList::MemberMap nativeMembers(ASTNode const* _currentScope) const override;
@@ -1622,6 +1632,7 @@ public:
 	u256 storageSize() const override;
 	bool hasSimpleZeroValueInMemory() const override { solAssert(false, ""); }
 	std::string richIdentifier() const override;
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool operator==(Type const& _other) const override;
 	std::string toString(bool _withoutDataLocation) const override;
 protected:
@@ -1646,6 +1657,7 @@ public:
 	std::string richIdentifier() const override;
 	bool operator==(Type const& _other) const override;
 	bool canBeStored() const override { return false; }
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool hasSimpleZeroValueInMemory() const override { solAssert(false, ""); }
 	MemberList::MemberMap nativeMembers(ASTNode const*) const override;
 
@@ -1685,6 +1697,7 @@ public:
 	std::string richIdentifier() const override;
 	bool operator==(Type const& _other) const override;
 	bool canBeStored() const override { return false; }
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool hasSimpleZeroValueInMemory() const override { solAssert(false, ""); }
 	MemberList::MemberMap nativeMembers(ASTNode const*) const override;
 
@@ -1718,6 +1731,7 @@ public:
 	unsigned calldataEncodedSize(bool) const override { return 32; }
 	bool canBeStored() const override { return false; }
 	bool isValueType() const override { return true; }
+	std::string eip712TypeName() const override { solAssert(false, ""); }
 	bool hasSimpleZeroValueInMemory() const override { solAssert(false, ""); }
 	std::string toString(bool) const override { return "inaccessible dynamic type"; }
 	Type const* decodingType() const override;
diff --git a/libsolidity/codegen/ir/IRGeneratorForStatements.cpp b/libsolidity/codegen/ir/IRGeneratorForStatements.cpp
index 6240d5a22..ae251fb7d 100644
--- a/libsolidity/codegen/ir/IRGeneratorForStatements.cpp
+++ b/libsolidity/codegen/ir/IRGeneratorForStatements.cpp
@@ -1944,9 +1944,9 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 		else if (member == "typehash")
 		{
 			Type const* arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
-			solAssert(!arg);
+			solAssert(!!arg);
 			StructType const* structType = dynamic_cast<StructType const*>(arg);
-			solAssert(!structType);
+			solAssert(!!structType);
 			StructDefinition const& struct_ = structType->structDefinition();
 			define(_memberAccess) << "0x" << struct_.typehash() << "\n";
 		}
```

Reply:
Note, in my `diff` `Enum` types get encoded as `uint8` and for:

> Definition: The atomic types are bytes1 to bytes32, uint8 to uint256, int8 to int256, bool and address. These correspond to their definition in Solidity. Note that there are no aliases uint and int. Note that contract addresses are always plain address. Fixed point numbers are not supported by the standard. Future versions of this standard may add new atomic types.

`eip712TypeName()` might possibly need to be modified or to throw an exception depending on the type.